File: prototype/src/contracts.ts
/** Supported Workspace surface. No queue, dispatcher, socket or allocation APIs. */
export interface Envelope {
  readonly version: number;
  readonly source_id: string;
  readonly source_description?: string;
  readonly time: string;
  readonly event_type: string;
  readonly workspace: readonly string[];
}

export interface ReadonlyPayload {
  readonly byteLength: number;
  at(index: number): number | undefined;
  toHex(): string;
  toUtf8(): string;
}

export interface IPayloadLease {
  readonly messageId: number;
  readonly envelope: Envelope;
  readonly payload: ReadonlyPayload;
}

export interface IWorkspaceMessageContext {
  Checkout(): IPayloadLease;
  /** false means this context already returned that lease; no second decrement. */
  Checkin(lease: IPayloadLease): boolean;
}

export interface IWorkspace {
  readonly name: string;
  Process(context: IWorkspaceMessageContext): void | Promise<void>;
}

export type FieldValue = string | number | boolean | null;
export interface RecordInput {
  readonly workspace: string;
  readonly fields: Readonly<Record<string, FieldValue>>;
}
export interface StoredRecord extends RecordInput { readonly id: number }
export interface IRecordStore { Append(record: RecordInput): void }
export interface RecordQuery {
  readonly workspaces?: readonly string[];
  readonly offset?: number;
  readonly limit?: number;
}
export interface IRecordQuery { Query(query: RecordQuery): readonly StoredRecord[] }
export interface IRecordExporter { Export(query: RecordQuery): string }

export interface ErrorBody {
  readonly code: string;
  readonly component: string;
  readonly source_id?: string;
  readonly workspace?: string;
  readonly detail?: string;
}
export interface ErrorAggregate {
  readonly body: ErrorBody;
  readonly first_seen: string;
  readonly last_seen: string;
  readonly count: number;
}
export interface IErrorSink { Report(body: ErrorBody, time?: string): void }
export interface IEchoSink { Emit(event: Readonly<Record<string, unknown>>): void }
export interface WorkspaceServices {
  readonly records: IRecordStore;
  readonly errors: IErrorSink;
  readonly echo: IEchoSink;
}
export interface WorkspacePlugin {
  readonly apiVersion: 1;
  createWorkspace(services: WorkspaceServices): IWorkspace;
}

File: src/Dsn.Contracts/Contracts.cs
using System.Text.Json;

namespace Dsn.Contracts;

public sealed record Envelope(int Version, string SourceId, JsonElement? SourceDescription,
    string Time, string EventType, IReadOnlyList<string> Workspace);

// The façade never exposes Memory/Span over the owned buffer. Every access checks the lease.
public interface IReadOnlyPayload
{
    int Length { get; }
    byte At(int index);
    byte[] Copy();
    string ToUtf8();
    string ToHex();
}
public interface IPayloadLease
{
    string MessageId { get; }
    Envelope Envelope { get; }
    IReadOnlyPayload Payload { get; }
}
public interface IMessageContext
{
    IPayloadLease Checkout();
    bool Checkin(IPayloadLease lease);
}
public interface IWorkspace
{
    string Name { get; }
    ValueTask ProcessAsync(IMessageContext message, CancellationToken cancellationToken);
}
public interface IWorkspacePlugin
{
    int ApiVersion { get; }
    IWorkspace Create(WorkspaceServices services);
}
public sealed record WorkspaceServices(IRecordStore Records, IErrorSink Errors);
public sealed record RecordInput(string Workspace, IReadOnlyDictionary<string, JsonElement> Fields);
public sealed record StoredRecord(long Id, string Workspace, IReadOnlyDictionary<string, JsonElement> Fields);
public sealed record RecordQuery(string[]? Workspaces = null, long AfterId = 0, int Limit = 100);
public interface IRecordStore { ValueTask AppendAsync(RecordInput record, CancellationToken cancellationToken = default); }
public interface IRecordQuery
{
    IReadOnlyList<StoredRecord> Query(RecordQuery query);
    IReadOnlyDictionary<string, string[]> Fields();
}
public interface IRecordExporter { string Export(RecordQuery query); }
public sealed record ErrorBody(string Code, string Component, string? SourceId = null, string? Workspace = null, string? Detail = null);
public sealed record ErrorAggregate(ErrorBody Body, DateTimeOffset FirstSeen, DateTimeOffset LastSeen, long Count);
public interface IErrorSink { void Report(ErrorBody body); }
public static class Fields
{
    public static IReadOnlyDictionary<string, JsonElement> From(params (string Name, object? Value)[] fields) =>
        fields.ToDictionary(p => p.Name, p => JsonSerializer.SerializeToElement(p.Value));
}
